storm/src/main/storm/mesos/MesosNimbus.java

```diff
@@ -577,288 +409,265 @@ public void assignSlots(Topologies topologies, Map<String, Collection<WorkerSlot
-      //    2. If the function is invoked a second time for the same slot, then it's either because the task
-      //       was finished or killed. Either way, the old offerId we used to launch the worker is invalid
-      if ((offerId == null || offer == null) && taskIDtoOfferMap.containsKey(taskId)) {
-        offer = taskIDtoOfferMap.get(taskId);
-        if (offer != null) {
-          offerId = offer.getId();
-          usingExistingOffer = true;
+  public Map<String, List<TaskInfo>> getTasksToLaunch(Topologies topologies,
+                                                      Map<String, Collection<WorkerSlot>> slots,
+                                                      Map<String, OfferResources> offerResourcesPerNode) {
+    Map<String, List<TaskInfo>> tasksToLaunchPerNode = new HashMap<>();
+
+    for (String topologyId : slots.keySet()) {
+      Collection<WorkerSlot> slotList = slots.get(topologyId);
+      TopologyDetails topologyDetails = topologies.getById(topologyId);
+      Set<String> hostsUsedSoFar = new HashSet<>();
+
+      double workerCpu = MesosCommon.topologyWorkerCpu(mesosStormConf, topologyDetails);
+      double workerMem = MesosCommon.topologyWorkerMem(mesosStormConf, topologyDetails);
+      double executorCpu = MesosCommon.executorCpu(mesosStormConf);
+      double executorMem = MesosCommon.executorMem(mesosStormConf);
+      double requiredCpu = workerCpu;
+      double requiredMem = workerMem;
+
+      for (WorkerSlot slot : slotList) {
+        String workerHost = slot.getNodeId();
+        Long workerPort = Long.valueOf(slot.getPort());
+
+        Boolean subtractExecutorResources = !hostsUsedSoFar.contains(workerHost);
+        hostsUsedSoFar.add(workerHost);
+
+        OfferResources offerResources = offerResourcesPerNode.get(slot.getNodeId());
+        String workerPrefix = "";
+        if (mesosStormConf.containsKey(MesosCommon.WORKER_NAME_PREFIX)) {
+          workerPrefix = MesosCommon.getWorkerPrefix(mesosStormConf, topologyDetails);
         }
-      }
 
-      // The following way of finding resources for the slot is bad!
-      // Suppose we have the following offers:
-      //     o1 - { host: h1 ports : 31000-32000 }
-      //     o2 - { host: h1 mem : 30000 cpu : 24 }
-      // because the offers are fragmented, they are useless
-      //   1. At this point a worker slot "h1-3000" associated with o1 for instance wouldn't
-      //      be launched because it doesn't have mem and cpu
-      //   2. o2 is useless because it doesn't have any ports and therefore won't be used at all
-      if (offerId != null && offer != null) {
-        // The fact that we are here means that this offer has a port. We need to find if it also has
-        // enough memory and cpu to launch the worker.
-        if (!toLaunch.containsKey(offerId)) {
-          toLaunch.put(offerId, new ArrayList<LaunchTask>());
+        if (hostsUsedSoFar.contains(workerHost)) {
```

Comment:
per our trio-coding/reviewing session, let's rewrite this stuff:

```
...
        // Account for executor resources only the 1st time we see this host for this topology.
        if (!hostsUsedSoFar.contains(workerHost)) {
          requiredCpu += executorCpu;
          requiredMem += executorMem;
        }

...
        if (!offerResources.isFit(mesosStormConf, topologyDetails, workerPort, hostsUsedSoFar.contains(workerHost)) {
          LOG.error(String.format("Unable to launch worker %s. Required cpu: %f, Required mem: %f. Available OfferResources : %s",
                                  workerHost, requiredCpu, requiredMem, offerResources));
          continue;
        }
...
        try {
...

          if (hostsUsedSoFar.contains(workerHost)) {
            // Need to account for resources, since this might be the 1st executor on this host for this topology (no way to tell right now)
          } else {
            // Oh, need to actually account for the resources
          }


        ...

        hostsUsedSoFar.add(workerHost);
```